Services/RequestServer/WebSocketImplCurl.cpp

```diff
@@ -0,0 +1,175 @@
+        // curl will discard headers with empty values unless we pass the header name followed by a semicolon.
+        ByteString header_string;
+        if (value.is_empty())
+            header_string = ByteString::formatted("{};", name);
+        else
+            header_string = ByteString::formatted("{}: {}", name, value);
+        curl_headers = curl_slist_append(curl_headers, header_string.characters());
+    }
+
+    if (auto const& protocols = info.protocols(); !protocols.is_empty()) {
+        StringBuilder protocol_builder;
+        protocol_builder.append("Sec-WebSocket-Protocol: "sv);
+        protocol_builder.append(ByteString::join(","sv, protocols));
+        curl_headers = curl_slist_append(curl_headers, protocol_builder.to_byte_string().characters());
+    }
+
+    if (auto const& extensions = info.extensions(); !extensions.is_empty()) {
+        StringBuilder protocol_builder;
+        protocol_builder.append("Sec-WebSocket-Extensions: "sv);
+        protocol_builder.append(ByteString::join(","sv, extensions));
+        curl_headers = curl_slist_append(curl_headers, protocol_builder.to_byte_string().characters());
+    }
+
+    set_option(CURLOPT_HTTPHEADER, curl_headers);
+    m_curl_string_lists.append(curl_headers);
+
+    CURLMcode const err = curl_multi_add_handle(m_multi_handle, m_easy_handle);
+    VERIFY(err == CURLM_OK);
+}
+
+bool WebSocketImplCurl::can_read_line()
+{
+    VERIFY_NOT_REACHED();
+}
+
+ErrorOr<ByteBuffer> WebSocketImplCurl::read(int max_size)
+{
+    auto buffer = TRY(ByteBuffer::create_uninitialized(max_size));
+    auto const read_bytes = TRY(m_read_buffer.read_some(buffer));
+    return buffer.slice(0, read_bytes.size());
+}
+
+ErrorOr<ByteString> WebSocketImplCurl::read_line(size_t)
+{
+    VERIFY_NOT_REACHED();
+}
+
+bool WebSocketImplCurl::send(ReadonlyBytes bytes)
+{
+    size_t sent = 0;
+    CURLcode const result = curl_ws_send(m_easy_handle, bytes.data(), bytes.size(), &sent, 0, 0);
+    (void)sent;
+
+    return result == CURLE_OK;
+}
+
+bool WebSocketImplCurl::eof()
+{
+    return m_read_buffer.is_eof();
+}
+
+void WebSocketImplCurl::discard_connection()
+{
+    if (m_read_notifier) {
+        m_read_notifier->close();
+        m_read_notifier = nullptr;
+    }
+    if (m_error_notifier) {
+        m_error_notifier->close();
+        m_error_notifier = nullptr;
+    }
+    curl_easy_cleanup(m_easy_handle);
+    m_easy_handle = nullptr;
+}
+
+void WebSocketImplCurl::did_connect()
+{
+    curl_socket_t socket_fd = CURL_SOCKET_BAD;
+    auto res = curl_easy_getinfo(m_easy_handle, CURLINFO_ACTIVESOCKET, &socket_fd);
+    VERIFY(res == CURLE_OK && socket_fd != CURL_SOCKET_BAD);
+
+    m_read_notifier = Core::Notifier::construct(socket_fd, Core::Notifier::Type::Read);
+    m_read_notifier->on_activation = [this] {
+        u8 buffer[65536];
+        size_t nread = 0;
+        CURLcode const result = curl_easy_recv(m_easy_handle, buffer, sizeof(buffer), &nread);
```

Comment:
curl_ws_recv in CURL_WS_RAW mode still doesn't pass us the header values, so we call up to curl_easy_recv here instead. Once we get rid of header parsing in LibWebSocket, we can call the "correct" method here.